src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -240,149 +238,326 @@ private static ImmutableArray<TextChangeRange> Merge(IReadOnlyList<ImmutableArra
-                        {
-                            // delete as much from old change as new change can
-                            // a new change deletion is a reduction in the old change insertion
-                            var oldChangeReduction = Math.Min(oldChange.NewLength, newChange.Span.Length);
-                            AddRange(list, new TextChangeRange(oldChange.Span, oldChange.NewLength - oldChangeReduction));
-                            oldDelta = oldDelta - oldChange.Span.Length + (oldChange.NewLength - oldChangeReduction);
-                            oldIndex++;
-
-                            // deduct the amount removed from oldChange from newChange's deletion span (since its already been applied)
-                            newChange = new TextChangeRange(new TextSpan(oldChange.Span.Start + oldDelta, newChange.Span.Length - oldChangeReduction), newChange.NewLength);
-                            goto nextOldChange;
-                        }
+
+                        var newDeletion = newChange.SpanLength + oldChange.Span.Length - oldChange.NewLength;
+                        newChange = new UnadjustedNewChange(oldChange.Span.Start + oldDelta, newDeletion, newChange.NewLength);
+                        if (tryGetNextOldChange(oldChanges, ref oldIndex, out oldChange)) continue;
+                        else break;
                     }
                 }
-                else
-                {
-                    // no more old changes, just add adjusted new change
-                    var adjustedNewChange = new TextChangeRange(new TextSpan(newChange.Span.Start - oldDelta, newChange.Span.Length), newChange.NewLength);
-                    AddRange(list, adjustedNewChange);
-                    newIndex++;
-                    goto nextNewChange;
-                }
             }
-            else
+
+            // there may be remaining old changes or remaining new changes (not both, and not neither)
+            switch (oldIndex == oldChanges.Length, newIndex == newChanges.Length)
+            {
+                case (true, true):
+                case (false, false):
+                    throw new InvalidOperationException();
+            }
+
+            while (oldIndex < oldChanges.Length)
+            {
+                addAndAdjustOldDelta(builder, ref oldDelta, oldChange);
+                tryGetNextOldChange(oldChanges, ref oldIndex, out oldChange);
+            }
+
+            while (newIndex < newChanges.Length)
             {
-                // no more new changes, just add remaining old changes
-                while (oldIndex < oldChanges.Length)
+                addAdjustedNewChange(builder, oldDelta, newChange);
+                tryGetNextNewChange(newChanges, ref newIndex, out newChange);
+            }
+
+            return builder.ToImmutableAndFree();
+
+            static bool tryGetNextOldChange(ImmutableArray<TextChangeRange> changes, ref int index, out TextChangeRange next)
```

Comment:
i don't think there's value in the two functions *if* you're going to still have to pass all the data to it explicitly.  either have one helper and pass the data you need to it.  Or have helpers, but have them be like:

```suggestion
            bool tryGetNextOldChange() => tryGetNextChange(oldChanges, ref oldIndex, out oldChange);
```

Reply:
Done, thanks!